### ISSUE TITLE
[Bug]: axis_angle_to_rotation_matrix low-angle branch is off by theta**3 / 6, 1.3e-10 at theta = 1e-3 in float64, and disagrees with the quaternion route by the same amount

### ISSUE THREAD

**Issue by Nicholas022400701** [kornia/kornia#4838]
### What happened?

`kornia.geometry.conversions.axis_angle_to_rotation_matrix` switches to a low-angle branch at `theta**2 <= 1e-6` (`theta <= 1e-3`, in every dtype). That branch returns `I + [v]x + [v]x**2 / 2`, Rodrigues' formula with `sin(theta) / theta` and `(1 - cos(theta)) / theta**2` replaced by `1` and `1 / 2`. The truncation error is `theta**3 / 6` per entry, so in float64 the branch is about a million ulp off next to a general branch that is exact to `2.2e-16`:

| `theta` (float64, generic axis) | branch | `\|R - exact rotation\|` | `\|R - quaternion route\|` | `det(R) - 1` |
| --- | --- | --- | --- | --- |
| `1.1e-3` | general | `1.1e-16` | `1.1e-16` | `-1.1e-16` |
| `9.99e-4` | low-angle | `1.3e-10` | `1.3e-10` | `2.5e-13` |
| `5e-4` | low-angle | `1.7e-11` | `1.7e-11` | `1.6e-14` |
| `1e-4` | low-angle | `1.3e-13` | `1.3e-13` | `2.2e-16` |
| `1e-5` | low-angle | `1.3e-16` | `1.3e-16` | `0` |

"quaternion route" is `quaternion_to_rotation_matrix(axis_angle_to_quaternion(v))`, which is exact on the whole range. `test_convention_both_branches_are_orthogonal_3947` (from #3989) requires the general branch to agree with that route to `1e-12`; the same assertion on the low-angle branch fails by a factor of 130. The result is also discontinuous in `theta` at the switch: `1.3e-10` of error on one side of `1e-3`, `1.1e-16` on the other. Round trips suffer the same way, `rotation_matrix_to_axis_angle(axis_angle_to_rotation_matrix(v))` is `1.3e-10` from `v` at `theta = 9.99e-4`, while `rotation_matrix_to_axis_angle` applied to the exact rotation of `v` returns `v` to `1.1e-19`. `So3.exp` in `kornia.geometry.liegroup` does keep the next series term (`0.5 - theta**2 / 48`) and uses a dtype-dependent switch, so this function is the odd one out.

In float32 the truncation error (`1.7e-10`) is below rounding (`3e-8` at these angles), so float32 users see nothing; this only costs float64 users, and it costs them in the range of small pose updates that float64 refinement loops live in.

### Minimal reproduction

```python
import torch
import kornia.geometry.conversions as C

axis = torch.tensor([1.0, 2.0, 3.0], dtype=torch.float64) / 14.0**0.5
for theta in (1.1e-3, 9.99e-4, 5e-4, 1e-4, 1e-5):
    v = (theta * axis)[None]
    direct = C.axis_angle_to_rotation_matrix(v)
    quat_route = C.quaternion_to_rotation_matrix(C.axis_angle_to_quaternion(v))
    branch = "general" if theta * theta > 1e-6 else "low-angle"
    print(
        f"theta={theta:<8} {branch:<9} |direct - quat_route| = {(direct - quat_route).abs().max().item():.2e}"
        f"  det(direct) - 1 = {torch.linalg.det(direct).item() - 1:.2e}"
    )
```

Output on `main` (`320f461`):

```
theta=0.0011   general   |direct - quat_route| = 1.11e-16  det(direct) - 1 = -1.11e-16
theta=0.000999 low-angle |direct - quat_route| = 1.33e-10  det(direct) - 1 = 2.49e-13
theta=0.0005   low-angle |direct - quat_route| = 1.67e-11  det(direct) - 1 = 1.55e-14
theta=0.0001   low-angle |direct - quat_route| = 1.34e-13  det(direct) - 1 = 2.22e-16
theta=1e-05    low-angle |direct - quat_route| = 1.34e-16  det(direct) - 1 = 0.00e+00
```

### Environment

```text
kornia 0.9.0rc1 (main @ 320f461) | torch 2.14.0+cpu | python 3.13.14 | OS Linux x86_64 | CUDA/ROCm None (available=False, GPU=none) | MPS=False
```

### Anything else?

Fix: keep the branch and the `1e-6` switch, and carry the `theta**2` term of each coefficient, `sin(theta) / theta = 1 - theta**2 / 6` and `(1 - cos(theta)) / theta**2 = 1 / 2 - theta**2 / 24`. The entries are then within `theta**5 / 120` of the exact rotation (`8.4e-18` at `theta = 1e-3`) and `det = 1 - theta**6 / 72`, so the branch is exact to float64 rounding and continuous with the general branch. Same output on the repro after the change:

```
theta=0.000999 low-angle |direct - quat_route| = 6.72e-18  det(direct) - 1 = -1.11e-16
theta=0.0005   low-angle |direct - quat_route| = 2.17e-19  det(direct) - 1 = 0.00e+00
theta=0.0001   low-angle |direct - quat_route| = 1.36e-20  det(direct) - 1 = 0.00e+00
```

What it changes elsewhere, measured on 20000 random axis-angles with `theta` in `[1e-6, 1e-3]`:

- float32: 2476 of 20000 matrices move, all with `theta > 4.2e-4`, by at most 4 ulp per entry; the maximum error against the exact rotation stays at `2.98e-8` before and after (that is float32 rounding). float16: bit-identical. The gradient at `theta = 0` stays finite (the branch is a polynomial in `v`).
- `tests/geometry/test_conversions.py`: 472 passed, 5 xfailed with the change (same as `main`). The new `test_low_angle_branch_agrees_with_the_quaternion_route` fails on `main` with `1.3323605174871134e-10 < 1e-15` and passes with the change; it checks the low-angle branch against Rodrigues' formula evaluated with `cos` and `sin` (`1 - cos(theta)` written as `2 sin(theta / 2)**2`) and against the quaternion route, on the same generic axis the #3947 test uses. The two comment lines in `test_convention_both_branches_are_orthogonal_3947` that quote the old branch (`1 + 2.5e-13`, `[[0.9999995, -0.001, 0.0], ...]`) are updated to the new values.

Fragment I would ship as `changelog.d/<pr>.fixed.md` (tell me if you want it filed as `breaking` because float32 outputs move by a few ulp):

> `axis_angle_to_rotation_matrix` now evaluates its low-angle branch (`theta**2 <= 1e-6`) with the series of `sin(theta) / theta` and `(1 - cos(theta)) / theta**2` through `theta**2`. The branch used to truncate Rodrigues' formula to `I + [v]x + [v]x**2 / 2`, an error of `theta**3 / 6` per entry: at `theta = 1e-3` in float64 the result was `1.3e-10` away from the rotation of the input and from `quaternion_to_rotation_matrix(axis_angle_to_quaternion(v))`, with `det = 1 + 2.5e-13`; it is now within `2.2e-16` of both with `det = 1` to rounding. In float32 about one input in eight with `theta` between `4e-4` and `1e-3` moves by up to 4 ulp with no change in the error against the exact rotation; float16 results are bit-identical.

Is this the direction you want, or would you rather move the switch itself (a dtype-dependent threshold like `So3.exp`, which would also need `1 - cos(theta)` rewritten as `2 sin(theta / 2)**2` in the general branch to be usable at small `theta` in float32)? I would keep the change to the two coefficients.

### Patch

The complete change is below so that nobody has to redo it; I will open it as the PR as soon as a maintainer confirms the direction.

<details>
<summary>2 files changed, 66 insertions(+), 21 deletions(-)</summary>

```diff
diff --git a/kornia/geometry/conversions.py b/kornia/geometry/conversions.py
index 5370465..0db6e31 100644
--- a/kornia/geometry/conversions.py
+++ b/kornia/geometry/conversions.py
@@ -462,27 +462,30 @@ def axis_angle_to_rotation_matrix(axis_angle: torch.Tensor) -> torch.Tensor:
     def _compute_rotation_matrix_taylor(axis_angle: torch.Tensor) -> torch.Tensor:
         rx, ry, rz = axis_angle.unbind(-1)
         k_one = torch.ones_like(rx)
-        k_half = 0.5 * k_one
 
         rx2, ry2, rz2 = rx * rx, ry * ry, rz * rz
         rxry, rxrz, ryrz = rx * ry, rx * rz, ry * rz
-
-        # second-order Taylor expansion of Rodrigues' formula:
-        #   R = I + [v]x + [v]x^2 / 2
-        # the first-order truncation had det = 1 + theta^2; the second-order
-        # truncation has det = 1 + theta^4 / 4, so the matrix is a rotation to
-        # the working precision across the whole low-angle branch
+        theta2 = rx2 + ry2 + rz2
+
+        # Rodrigues' formula R = I + (sin(theta) / theta) [v]x + ((1 - cos(theta)) / theta^2) [v]x^2 with each
+        # coefficient replaced by its series through theta^2: sin(theta) / theta = 1 - theta^2 / 6 + O(theta^4) and
+        # (1 - cos(theta)) / theta^2 = 1 / 2 - theta^2 / 24 + O(theta^4). The entries are then within theta^5 / 120
+        # of the exact rotation (8.4e-18 at theta = 1e-3, the top of this branch) and det = 1 - theta^6 / 72, so
+        # the branch is exact to float64 rounding. Keeping only the leading terms, R = I + [v]x + [v]x^2 / 2,
+        # left an error of theta^3 / 6 (1.7e-10 at theta = 1e-3) and det = 1 + theta^4 / 4.
+        k_sin = k_one - theta2 / 6.0  # sin(theta) / theta
+        k_cos = 0.5 * k_one - theta2 / 24.0  # (1 - cos(theta)) / theta^2
         return torch.stack(
             [
-                k_one - k_half * (ry2 + rz2),
-                -rz + k_half * rxry,
-                ry + k_half * rxrz,
-                rz + k_half * rxry,
-                k_one - k_half * (rx2 + rz2),
-                -rx + k_half * ryrz,
-                -ry + k_half * rxrz,
-                rx + k_half * ryrz,
-                k_one - k_half * (rx2 + ry2),
+                k_one - k_cos * (ry2 + rz2),
+                -k_sin * rz + k_cos * rxry,
+                k_sin * ry + k_cos * rxrz,
+                k_sin * rz + k_cos * rxry,
+                k_one - k_cos * (rx2 + rz2),
+                -k_sin * rx + k_cos * ryrz,
+                -k_sin * ry + k_cos * rxrz,
+                k_sin * rx + k_cos * ryrz,
+                k_one - k_cos * (rx2 + ry2),
             ],
             dim=-1,
         ).reshape(list(axis_angle.shape[:-1]) + [3, 3])
diff --git a/tests/geometry/test_conversions.py b/tests/geometry/test_conversions.py
index 7f1c2c3..f267114 100644
--- a/tests/geometry/test_conversions.py
+++ b/tests/geometry/test_conversions.py
@@ -2235,9 +2235,10 @@ class TestAngleAxisToRotationMatrix(BaseTester):
         #       det = 0.9999974535249636 and max|R @ R.T - I| = 2.5464750363912714e-06;
         #   (2) theta = 1e-3 takes the low-angle branch, which returned the first-order Taylor
         #       matrix [[1, -rz, ry], [rz, 1, -rx], [-ry, rx, 1]] with det = 1 + theta**2 = 1.000001.
-        # The low-angle branch now returns the second-order Taylor expansion R = I + [v]x + [v]x^2/2,
-        # whose determinant is 1 + theta**4 / 4 -- at theta = 1e-3 that is 1 + 2.5e-13, a rotation
-        # to the working precision -- so both branches must now pass the same orthogonality checks.
+        # The low-angle branch now evaluates Rodrigues' formula with sin(theta) / theta and
+        # (1 - cos(theta)) / theta**2 replaced by their series through theta**2, so its determinant is
+        # 1 - theta**6 / 72 (1.0 to float64 rounding at theta = 1e-3) and both branches must pass the same
+        # orthogonality checks.
         # The general branch must also agree with the quaternion route on a generic (non-axis
         # aligned) rotation, which is where the eps defect showed up as an axis-dependent error.
         # float64 is hardcoded and the dtype fixture dropped because both cells are float64 facts:
@@ -2250,8 +2251,9 @@ class TestAngleAxisToRotationMatrix(BaseTester):
         #   (R @ R.T - torch.eye(3, dtype=torch.float64)).abs().max()   -> 0.0
         #   t = torch.tensor([[0., 0., 1e-3]], dtype=torch.float64)   # theta**2 == 1e-06 exactly
         #   axis_angle_to_rotation_matrix(t)[0].tolist()
-        #     -> [[0.9999995, -0.001, 0.0], [0.001, 0.9999995, 0.0], [0.0, 0.0, 1.0]]
-        #   torch.linalg.det(that).item()                               -> 1.00000000000025
+        #     -> [[0.9999995000000417, -0.0009999998333333332, 0.0],
+        #         [0.0009999998333333332, 0.9999995000000417, 0.0], [0.0, 0.0, 1.0]]
+        #   torch.linalg.det(that).item()                               -> 1.0
         #   g = torch.tensor([[1., 2., 3.]], dtype=torch.float64) * 0.6 / math.sqrt(14.0)  # generic axis
         #   Rg = axis_angle_to_rotation_matrix(g)[0]
         #   torch.linalg.det(Rg).item()                                 -> 1.0
@@ -2287,6 +2289,46 @@ class TestAngleAxisToRotationMatrix(BaseTester):
             "kornia#3947: axis_angle_to_rotation_matrix disagrees with the quaternion route"
         )
 
+    def test_low_angle_branch_agrees_with_the_quaternion_route(self, device):
+        # The low-angle branch (theta**2 <= 1e-6) kept only the leading terms of Rodrigues' formula,
+        # R = I + [v]x + [v]x**2 / 2, whose error is theta**3 / 6: 1.66e-10 per entry at theta = 1e-3 in float64,
+        # next to a general branch that is exact to 2.2e-16. axis_angle_to_quaternion followed by
+        # quaternion_to_rotation_matrix was exact there all along, so the two routes disagreed by 1.66e-10,
+        # a hundred times the 1e-12 that test_convention_both_branches_are_orthogonal_3947 demands of the
+        # general branch. The coefficients now carry their theta**2 terms and the entries are within
+        # theta**5 / 120 of the exact rotation, the same 1e-12 holds on both branches, and the reference below
+        # is Rodrigues' formula evaluated with cos and sin (1 - cos(theta) written as 2 sin(theta / 2)**2 so that
+        # it does not cancel).
+        # Snippet used to generate expected (torch only, executed on cpu float64):
+        #   v = torch.tensor([[0.0, 0.0, 1e-3]], dtype=torch.float64)
+        #   axis_angle_to_rotation_matrix(v)[0, 0, 1].item()   -> -0.0009999998333333332 (was -0.001)
+        #   -math.sin(1e-3)                                      -> -0.0009999998333333417
+        _skip_if_dtype_unavailable(device, torch.float64)
+        axis = torch.tensor([1.0, 2.0, 3.0], dtype=torch.float64) / 14.0**0.5
+        theta = torch.tensor([9.99e-4, 5e-4, 1e-4, 1e-5], dtype=torch.float64)
+        axis_angle = (theta[:, None] * axis).to(device)
+
+        skew = torch.tensor(
+            [[0.0, -axis[2], axis[1]], [axis[2], 0.0, -axis[0]], [-axis[1], axis[0], 0.0]], dtype=torch.float64
+        ).to(device)
+        theta = theta.to(device)
+        expected = (
+            torch.eye(3, device=device, dtype=torch.float64)
+            + theta.sin()[:, None, None] * skew
+            + (2.0 * (0.5 * theta).sin() ** 2)[:, None, None] * (skew @ skew)
+        )
+
+        rot = axis_angle_to_rotation_matrix(axis_angle)
+        assert (rot - expected).abs().max().item() < 1e-15, (
+            "the low-angle branch of axis_angle_to_rotation_matrix is not the rotation of its input"
+        )
+        quat_route = kornia.geometry.conversions.quaternion_to_rotation_matrix(
+            kornia.geometry.conversions.axis_angle_to_quaternion(axis_angle)
+        )
+        assert (rot - quat_route).abs().max().item() < 1e-12, (
+            "the low-angle branch of axis_angle_to_rotation_matrix disagrees with the quaternion route"
+        )
+
     def test_convention_accepts_any_leading_batch_dimensions_3955(self, device):
         # Convention: axis_angle_to_rotation_matrix accepts (*, 3) and returns (*, 3, 3) -- what its
         # own shape guard promises in the message it raises ("Input size must be a (*, 3) tensor")
```

</details>

AI disclosure: I used an AI coding agent to help find this and write this report, the patch and the tests. I have checked the numbers myself and I will answer questions personally.
